cpp/src/fil/fil.cu

```diff
@@ -173,8 +203,21 @@ struct forest {
 
   virtual void infer(predict_params params, cudaStream_t stream) = 0;
 
-  void predict(
-    const raft::handle_t& h, float* preds, const float* data, size_t num_rows, bool predict_proba)
+  virtual void predict(const raft::handle_t& h,
```

Comment:
I'd strongly recommend we reconsider this structure. There is no reason that the downstream caller needs to have a single `predict` API with type erasure for both `float` and `double`. I also want to make sure we are absolutely confident on what the performance implications will be of making this a virtual function, particularly for applications that are making many repeated calls to `predict` in a short period.

Reply:
Done.